lib/test.rb

```diff
@@ -420,38 +420,69 @@ def setup_formulae_deps_instances(formula, formula_name)
         return
       end
 
-      build_dependents_from_source = %w[
+      build_dependents_from_source_whitelist = %w[
         cabal-install
         docbook-xsl
         ghc
         go
         ocaml
+        ocaml-findlib
+        ocaml-num
         openjdk
         rust
-      ].include?(formula_name)
+      ]
 
       uses_args = []
       uses_args << "--recursive" unless Homebrew.args.skip_recursive_dependents?
-      # whitelist specific formula where we want to test build dependants.
-      uses_args << "--include-build" if build_dependents_from_source
       dependents =
-        Utils.popen_read("brew", "uses", "--include-test", *uses_args, formula_name)
+        Utils.popen_read("brew", "uses", "--include-build", "--include-test", *uses_args, formula_name)
              .split("\n")
       dependents -= @formulae
       dependents = dependents.map { |d| Formulary.factory(d) }
 
-      if build_dependents_from_source
-        @source_dependents = dependents
-        @testable_dependents = @source_dependents.select(&:test_defined?)
-        @bottled_dependents = []
-        return
+      dependents = dependents.zip(dependents.map do |f|
+        if Homebrew.args.skip_recursive_dependents?
+          f.deps
+        else
+          f.recursive_dependencies
+        end.reject(&:optional?)
+      end)
+
+      # Defer formulae which could be tested later
+      # i.e. formulae that also depend on something else yet to be built in this test run.
+      dependents.select! do |_, deps|
+        still_to_test = @formulae - @formulae_that_have_been_built
+        (deps.map { |d| d.to_formula.full_name } & still_to_test).empty?
       end
 
-      @source_dependents = []
+      # Split into dependents that we could potentially be building from source and those
+      # we should not. The criteria is that it depends on a formula in the whitelist and
+      # that formula has been, or will be, built in this test run.
+      @source_dependents, dependents = dependents.partition do |_, deps|
+        deps.any? do |d|
+          full_name = d.to_formula.full_name
+
+          next false unless build_dependents_from_source_whitelist.include?(full_name)
+
+          @formulae.include?(full_name)
+        end
+      end
```

Comment:
We look at the dependents and split them into those that should be built from source and which should not. The check is to see if any of the dependent's dependencies (recursively if enabled) is on the whitelist _and_ that dependency is in the list of top-level formulae being tested.

This should be seen as a replacement of the existing build from source selection logic rather than just an addition. It allows much greater flexibility to defer builds and thus avoid repeating builds or polluting `etc` and `var`.

Otherwise, deferring would not be possible for mixed BFS & bottle cases and additional checks would need to be done at the defer stage to prevent that.